overleaf/real-time
# [ChannelManager] enforce a strict sequence of redis (un)subscribes

### Description

Chained onto #132 

This PR implements a strict sequence of subscribe and unsubscribe calls.

It is a safe guard for future work of arbitrary cleanup calls.

Subscribe waits for unsubscribe requests. Unsubscribe waits for
 subscribe requests. Unsubscribe breaks the chain.

Additionally I added shortcuts into the unsubscribe process to skip the redis request
 when we detect a race/overtaking request. This is important for the future delayed
 unsubscribe calls -- they should not block any new subscribes or do duplicate work.

1st subscribe -> 1st unsubscribe -> 2nd subscribe -> 2nd unsubscribe

We can skip the 1st unsubscribe, when the 2nd subscribe came in before the 1st subscribe completed. We can also skip the 1st unsubscribe when the 2nd unsubscribe came in before the 1st subscribe completed.

#### Review

I could not find a simple way to tell whether a Promise completed already, so there is a home grown solution which lets a given Promise race against another that completes 1ms later.
I am happy to replace it with something simpler. Moving to a new coffee-script version or decaff would enable `async function`s, but we are not there yet.

Regarding the acceptance tests:

`test/acceptance/coffee/PubSubRace.coffee -> when the client disconnects before joinDoc completes`:

Due to different code paths in socket.io for `joinProject` and `disconnect` requests from the client, we are very likely not getting into a race (two vs one process.nextTick calls).

#### Related Issues / PRs

https://github.com/overleaf/issues/issues/3002

#### Potential Impact

High. clients could be connected without getting any updates.

#### Manual Testing Performed

- added lots of unit tests
- a new acceptance test suite
- open lots of editor tabs, they should not see each other
